src/common/Throttle.h

```diff
@@ -330,4 +331,183 @@ class OrderedThrottle {
   uint32_t waiters = 0;
 };
 
+
+template <typename T, typename I>
+class TokenBucketThrottle {
+
+  struct Bucket {
+    CephContext *cct;
+    const std::string name;
+    std::atomic<int64_t> remain = { 0 }, max = { 0 };
+
+    Bucket(CephContext *cct, const std::string& n, int64_t m)
+      : cct(cct), name(n),
+	remain(m), max(m)
+    {
+    }
+
+    int64_t get(int64_t c) {
+      if (0 == max) {
+	return 0;
+      }
+
+      assert(c >= 0);
+
+      int64_t got = 0;
+      if (remain >= c) {
+	// There is enough token in bucket, take c.
+	got = c;
+	remain -= c;
+      } else {
+	// There is not enough, take all remained.
+	got = remain;
+	remain = 0;
+      }
+      return got;
+    }
+
+    int64_t put(int64_t c) {
+      if (0 == max) {
+	return 0;
+      }
+
+      assert(c >= 0);
+      if (c) {
+	// put c tokens into bucket
+	int64_t current = remain;
+	if ((current + c) <= (int64_t)max) {
+	  remain += c;
+	} else {
+	  remain = (int64_t)max;
+	}
+      }
+      return remain;
+    }
+
+    void set_max(int64_t m) {
+      if (remain > m)
+	remain = m;
+      max = m;
+    }
+  };
+
+  struct Blocker {
+    int64_t tokens_requested;
+    T *got_token_handler;
+    I *item;
+
+    Blocker(int64_t _tokens_requested, T *handler, I *_item)
+    : tokens_requested(_tokens_requested), got_token_handler(handler),
+      item(_item) {}
+  };
+
+  CephContext *m_cct;
+  Bucket m_throttle;
+  int64_t m_avg = 0;
+  SafeTimer *m_timer;
+  Mutex *m_timer_lock;
+  FunctionContext *m_token_ctx;
+  list<Blocker*> m_blockers;
+  Mutex m_lock;
+
+public:
+TokenBucketThrottle(
+    CephContext *cct,
+    int64_t capacity,
```

Comment:
Nit: I still don't understand the use of signed typed throughout. You either have capacity or you don't so you shouldn't have a "negatively" filled bucket.

Reply:
Yes, uint64_t should be better. thanx